### DISCUSSION TITLE
[Proposal] "Stable" C API

### DISCUSSION THREAD

**Discussion by Ronsor** [ggml-org/llama.cpp#177] [Ideas]
I propose refactoring `main.cpp` into a library (`llama.cpp`, compiled to `llama.so`/`llama.a`/whatever) and making `main.cpp` a simple driver program. A simple C API should be exposed to access the model, and then bindings can more easily be written for Python, node.js, or whatever other language.

This would partially solve #82 and #162.

Edit: on that note, is it possible to do inference from two or more prompts on different threads? If so, serving multiple people would be possible without multiple copies of model weights in RAM.

**Comment by bakkot** [comment]
For anyone wanting to do this, see an initial attempt in https://github.com/ggerganov/llama.cpp/pull/77, and in particular [this comment](https://github.com/ggerganov/llama.cpp/pull/77#pullrequestreview-1337544258) on ggerganov's preferred approach. Should be pretty straightforward I think.

**Comment by v3ss0n** [comment]
It is already ongoing , check PR #77

**Comment by ggerganov** [comment]
Yes, see the comment https://github.com/ggerganov/llama.cpp/pull/77#pullrequestreview-1337544258 as @bakkot suggested. This is the way 🦙

**Comment by v3ss0n** [comment]
ah @bakkot  beat me to it while i was writing. @Ronsor  please close this , and the project have Discussion now https://github.com/ggerganov/llama.cpp/discussions
